bindings/c/test/apitester/TesterTransactionExecutor.cpp

```diff
@@ -496,6 +501,215 @@ class AsyncTransactionContext : public TransactionContextBase {
+	                     std::shared_ptr<IDatabaseActor> dbActor,
+	                     TTaskFct cont,
+	                     IScheduler* scheduler)
+	  : DatabaseContextBase(db, dbActor, cont, scheduler) {}
+
+protected:
+	void doContinueAfter(fdb::Future f, TTaskFct cont) override {
+		std::unique_lock<std::mutex> lock(mutex);
+		if (isDone) {
+			return;
+		}
+		callbackMap[f] = DBCallbackInfo{ f, cont, shared_from_this(), timeNow() };
+		lock.unlock();
+		try {
+			f.then([this](fdb::Future f) { futureReadyCallback(f, this); });
+		} catch (std::runtime_error& err) {
+			lock.lock();
+			callbackMap.erase(f);
+			lock.unlock();
+			dbActor->complete(fdb::Error(error_code_operation_failed));
+		}
+	}
+
+	static void futureReadyCallback(fdb::Future f, void* param) {
+		try {
+			AsyncDatabaseContext* dbCtx = (AsyncDatabaseContext*)param;
+			dbCtx->onFutureReady(f);
+		} catch (std::runtime_error& err) {
+			fmt::print("Unexpected exception in callback {}\n", err.what());
+			abort();
+		} catch (...) {
+			fmt::print("Unknown error in callback\n");
+			abort();
+		}
+	}
+
+	void onFutureReady(fdb::Future f) {
+		auto endTime = timeNow();
+		injectRandomSleep();
+		// Hold a reference to this to avoid it to be
+		// destroyed before releasing the mutex
+		auto thisRef = shared_from_this();
+		std::unique_lock<std::mutex> lock(mutex);
+		auto iter = callbackMap.find(f);
+		ASSERT(iter != callbackMap.end());
+		DBCallbackInfo cbInfo = iter->second;
+		callbackMap.erase(iter);
+		if (isDone) {
+			return;
+		}
+		lock.unlock();
+		fdb::Error err = f.error();
+		auto waitTimeUs = timeElapsedInUs(cbInfo.startTime, endTime);
+		if (waitTimeUs > LONG_WAIT_TIME_US) {
+			fmt::print("Long waiting time on a future: {:.3f}s, return code {} ({})\n",
+			           microsecToSec(waitTimeUs),
+			           err.code(),
+			           err.what());
+		}
+
+		if (err.code() == error_code_success) {
+			scheduler->schedule(cbInfo.cont);
+		} else {
+			dbActor->complete(err);
+		}
+	}
+
+	// Inject a random sleep with a low probability
+	void injectRandomSleep() {
+		if (Random::get().randomBool(0.01)) {
+			std::this_thread::sleep_for(std::chrono::milliseconds(Random::get().randomInt(1, 5)));
+		}
+	}
+
+	void cleanUp() override {
+		DatabaseContextBase::cleanUp();
+
+		// Cancel all pending operations
+		// Note that the callbacks of the cancelled futures will still be called
+		std::unique_lock<std::mutex> lock(mutex);
+		std::vector<fdb::Future> futures;
+		for (auto& iter : callbackMap) {
+			futures.push_back(iter.second.future);
+		}
+		lock.unlock();
+		for (auto& f : futures) {
+			f.cancel();
+		}
+	}
+
+	// Object references for a future callback
+	struct DBCallbackInfo {
+		fdb::Future future;
+		TTaskFct cont;
+		std::shared_ptr<IDatabaseContext> thisRef;
+		TimePoint startTime;
+	};
+
+	// Map for keeping track of future waits and holding necessary object references
+	std::unordered_map<fdb::Future, DBCallbackInfo> callbackMap;
```

Comment:
I am wondering if we need support for multiple futures here. In TransactionContext we are bundling all operations that belong to one transaction, but here if there are no transactions, probably there is no reason for bundling multiple DB operations into one context. We could have one executeDBOperation call per each DB operation. Then we also do not need the explicit done() call, because the operation is simply done when its future is ready.

Reply:
You're right, yeah. I'll work on simplifying this further.